Source/Core/VideoBackends/Software/TransformUnit.cpp

```diff
@@ -208,42 +208,34 @@ static inline float SafeDivide(float n, float d)
 	return (d==0) ? (n>0?1:0) : n/d;
 }
 
-static void LightColor(const Vec3 &pos, const Vec3 &normal, u8 lightNum, const LitChannel &chan, Vec3 &lightCol)
+static inline float CalculateLightAttn(const LightPointer *light, Vec3 &ldir, const Vec3 &normal, const LitChannel &chan)
 {
-	const LightPointer *light = (const LightPointer*)&xfmem.lights[lightNum];
+	float attn = 1.0f;
 
-	if (!(chan.attnfunc & 1))
+	switch (chan.attnfunc)
 	{
-		// atten disabled
-		switch (chan.diffusefunc)
+		case LIGHTATTN_NONE:
+		case LIGHTATTN_DIR:
 		{
-			case LIGHTDIF_NONE:
-				AddIntegerColor(light->color, lightCol);
-				break;
-			case LIGHTDIF_SIGN:
-				{
-					Vec3 ldir = (light->pos - pos).Normalized();
-					float diffuse = ldir * normal;
-					AddScaledIntegerColor(light->color, diffuse, lightCol);
-				}
-				break;
-			case LIGHTDIF_CLAMP:
-				{
-					Vec3 ldir = (light->pos - pos).Normalized();
-					float diffuse = std::max(0.0f, ldir * normal);
-					AddScaledIntegerColor(light->color, diffuse, lightCol);
-				}
-				break;
-			default: _assert_(0);
+			ldir = ldir.Normalized();
+			if (ldir == Vec3(0.0f, 0.0f, 0.0f))
+				ldir = normal;
+			break;
 		}
-	}
-	else // spec and spot
-	{
-		// not sure about divide by zero checks
-		Vec3 ldir = light->pos - pos;
-		float attn;
-
-		if (chan.attnfunc == 3) // spot
+		case LIGHTATTN_SPEC:
+		{
+			ldir = ldir.Normalized();
+			attn = (ldir * normal) >= 0.0 ? std::max(0.0f, light->dir * normal) : 0;
```

Comment:
light->dir is simply the direction of the light.
ldir is the calculated vector between the light position and the model position for lighting calculation purposes.  I see a lot of people use ldir for this var name, but maybe a name like surfaceToLightDir would be more clear (but that's a long name lol)